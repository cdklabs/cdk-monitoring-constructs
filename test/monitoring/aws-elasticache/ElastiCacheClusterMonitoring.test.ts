import { Stack } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";

import {
  AlarmWithAnnotation,
  ElastiCacheClusterMonitoring,
  ElastiCacheClusterType,
} from "../../../lib";
import { TestMonitoringScope } from "../TestMonitoringScope";

const DummyMemcachedClusterId = "DummyMemcachedClusterId";
const DummyRedisClusterId = "DummyRedisClusterId";

test("snapshot test: no alarms", () => {
  const stack = new Stack();

  const scope = new TestMonitoringScope(stack, "Scope");

  new ElastiCacheClusterMonitoring(scope, {
    clusterType: ElastiCacheClusterType.MEMCACHED,
  });

  new ElastiCacheClusterMonitoring(scope, {
    clusterType: ElastiCacheClusterType.REDIS,
  });

  expect(Template.fromStack(stack)).toMatchSnapshot();
});

test("snapshot test: all alarms", () => {
  const stack = new Stack();

  const scope = new TestMonitoringScope(stack, "Scope");

  let numAlarmsCreatedForMemcached = 0;
  let numAlarmsCreatedForRedis = 0;

  new ElastiCacheClusterMonitoring(scope, {
    clusterType: ElastiCacheClusterType.MEMCACHED,
    addCpuUsageAlarm: { Warning: { maxUsagePercent: 10 } },
    addMaxItemsCountAlarm: { Warning: { maxItemsCount: 20 } },
    addMaxEvictedItemsCountAlarm: { Warning: { maxItemsCount: 30 } },
    addMinFreeableMemoryAlarm: { Warning: { minFreeableMemoryInBytes: 40 } },
    addMaxUsedSwapMemoryAlarm: { Warning: { maxUsedSwapMemoryInBytes: 50 } },
    useCreatedAlarms: {
      consume(alarms: AlarmWithAnnotation[]) {
        numAlarmsCreatedForMemcached = alarms.length;
      },
    },
  });

  new ElastiCacheClusterMonitoring(scope, {
    clusterType: ElastiCacheClusterType.REDIS,
    addCpuUsageAlarm: { Warning: { maxUsagePercent: 11 } },
    addRedisEngineCpuUsageAlarm: { Warning: { maxUsagePercent: 10 } },
    addMaxItemsCountAlarm: { Warning: { maxItemsCount: 21 } },
    addMaxEvictedItemsCountAlarm: { Warning: { maxItemsCount: 31 } },
    addMinFreeableMemoryAlarm: { Warning: { minFreeableMemoryInBytes: 41 } },
    addMaxUsedSwapMemoryAlarm: { Warning: { maxUsedSwapMemoryInBytes: 51 } },
    useCreatedAlarms: {
      consume(alarms: AlarmWithAnnotation[]) {
        numAlarmsCreatedForRedis = alarms.length;
      },
    },
  });

  expect(numAlarmsCreatedForMemcached).toStrictEqual(5);
  expect(numAlarmsCreatedForRedis).toStrictEqual(6);
  expect(Template.fromStack(stack)).toMatchSnapshot();
});

test("snapshot test: cluster ID specified", () => {
  const stack = new Stack();

  const scope = new TestMonitoringScope(stack, "Scope");

  new ElastiCacheClusterMonitoring(scope, {
    clusterType: ElastiCacheClusterType.MEMCACHED,
    clusterId: DummyMemcachedClusterId,
    humanReadableName: DummyMemcachedClusterId,
    alarmFriendlyName: DummyMemcachedClusterId,
  });

  new ElastiCacheClusterMonitoring(scope, {
    clusterType: ElastiCacheClusterType.REDIS,
    clusterId: DummyRedisClusterId,
    humanReadableName: DummyRedisClusterId,
    alarmFriendlyName: DummyRedisClusterId,
  });

  expect(Template.fromStack(stack)).toMatchSnapshot();
});

test("validation test: redisEngineCpuUsageAlarm added for non-redis cluster ", () => {
  const stack = new Stack();
  const scope = new TestMonitoringScope(stack, "Scope");

  expect(
    () =>
      new ElastiCacheClusterMonitoring(scope, {
        clusterType: ElastiCacheClusterType.MEMCACHED,
        addRedisEngineCpuUsageAlarm: { Warning: { maxUsagePercent: 10 } },
      }),
  ).toThrowError(
    "It is only possible to alarm on Redis Engine CPU Usage for Redis clusters",
  );
});
