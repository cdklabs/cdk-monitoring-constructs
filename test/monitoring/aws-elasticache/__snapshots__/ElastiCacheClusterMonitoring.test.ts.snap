// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test: all alarms 1`] = `
Object {
  "Resources": Object {
    "ScopeTestMemcachedALLCPUUsageWarningB9311F1B": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "The CPU usage is too high.",
        "AlarmName": "Test-Memcached-ALL-CPU-Usage-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "Cluster CPU Utilization",
            "MetricStat": Object {
              "Metric": Object {
                "MetricName": "CPUUtilization",
                "Namespace": "AWS/ElastiCache",
              },
              "Period": 300,
              "Stat": "Maximum",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 10,
        "TreatMissingData": "missing",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestMemcachedALLItemsCountWarningBCDA36CB": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "The number of items in the cache is too high.",
        "AlarmName": "Test-Memcached-ALL-Items-Count-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "Count",
            "MetricStat": Object {
              "Metric": Object {
                "MetricName": "CurrItems",
                "Namespace": "AWS/ElastiCache",
              },
              "Period": 300,
              "Stat": "Maximum",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 20,
        "TreatMissingData": "missing",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestMemcachedALLItemsEvictedWarning7D7F5BE2": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "The number of items evicted by the cache is too high.",
        "AlarmName": "Test-Memcached-ALL-Items-Evicted-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "Evictions",
            "MetricStat": Object {
              "Metric": Object {
                "MetricName": "Evictions",
                "Namespace": "AWS/ElastiCache",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 30,
        "TreatMissingData": "missing",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestMemcachedALLMemoryFreeableWarningDCD9279E": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "The size of freeable memory is too low.",
        "AlarmName": "Test-Memcached-ALL-Memory-Freeable-Warning",
        "ComparisonOperator": "LessThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "Freeable",
            "MetricStat": Object {
              "Metric": Object {
                "MetricName": "FreeableMemory",
                "Namespace": "AWS/ElastiCache",
              },
              "Period": 300,
              "Stat": "Average",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 40,
        "TreatMissingData": "missing",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestMemcachedALLMemorySwapWarning19BA1398": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "The size of swap memory used is too high.",
        "AlarmName": "Test-Memcached-ALL-Memory-Swap-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "Swap",
            "MetricStat": Object {
              "Metric": Object {
                "MetricName": "SwapUsage",
                "Namespace": "AWS/ElastiCache",
              },
              "Period": 300,
              "Stat": "Average",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 50,
        "TreatMissingData": "missing",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestRedisALLCPUUsageWarningF75B1746": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "The CPU usage is too high.",
        "AlarmName": "Test-Redis-ALL-CPU-Usage-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "Cluster CPU Utilization",
            "MetricStat": Object {
              "Metric": Object {
                "MetricName": "CPUUtilization",
                "Namespace": "AWS/ElastiCache",
              },
              "Period": 300,
              "Stat": "Maximum",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 11,
        "TreatMissingData": "missing",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestRedisALLItemsCountWarning95D20971": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "The number of items in the cache is too high.",
        "AlarmName": "Test-Redis-ALL-Items-Count-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "Count",
            "MetricStat": Object {
              "Metric": Object {
                "MetricName": "CurrItems",
                "Namespace": "AWS/ElastiCache",
              },
              "Period": 300,
              "Stat": "Maximum",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 21,
        "TreatMissingData": "missing",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestRedisALLItemsEvictedWarningF77927A9": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "The number of items evicted by the cache is too high.",
        "AlarmName": "Test-Redis-ALL-Items-Evicted-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "Evictions",
            "MetricStat": Object {
              "Metric": Object {
                "MetricName": "Evictions",
                "Namespace": "AWS/ElastiCache",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 31,
        "TreatMissingData": "missing",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestRedisALLMemoryFreeableWarning6141794B": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "The size of freeable memory is too low.",
        "AlarmName": "Test-Redis-ALL-Memory-Freeable-Warning",
        "ComparisonOperator": "LessThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "Freeable",
            "MetricStat": Object {
              "Metric": Object {
                "MetricName": "FreeableMemory",
                "Namespace": "AWS/ElastiCache",
              },
              "Period": 300,
              "Stat": "Average",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 41,
        "TreatMissingData": "missing",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestRedisALLMemorySwapWarningBE380050": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "The size of swap memory used is too high.",
        "AlarmName": "Test-Redis-ALL-Memory-Swap-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "Swap",
            "MetricStat": Object {
              "Metric": Object {
                "MetricName": "SwapUsage",
                "Namespace": "AWS/ElastiCache",
              },
              "Period": 300,
              "Stat": "Average",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 51,
        "TreatMissingData": "missing",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
  },
}
`;

exports[`snapshot test: cluster ID specified 1`] = `Object {}`;

exports[`snapshot test: no alarms 1`] = `Object {}`;
